python_modules/dagster/dagster/_core/definitions/data_version.py

```diff
@@ -581,23 +580,23 @@ def _get_stale_causes_materialized(self, key: "AssetKeyPartitionKey") -> Iterato
 
     @cached_method
     def _get_stale_root_causes(self, key: "AssetKeyPartitionKey") -> Sequence[StaleCause]:
-        causes = self._get_stale_causes(key=key)
-        root_pairs = sorted([pair for cause in causes for pair in self._gather_leaves(cause)])
-        # After sorting the pairs, we can drop the level and de-dup using an
-        # ordered dict as an ordered set. This will give us unique root causes,
-        # sorted by level.
-        roots: Dict[StaleCause, None] = OrderedDict()
-        for root_cause in [leaf_cause for _, leaf_cause in root_pairs]:
-            roots[root_cause] = None
-        return list(roots.keys())
-
-    # The leaves of the cause tree for an asset are the root causes of its staleness.
-    def _gather_leaves(self, cause: StaleCause, level: int = 0) -> Iterator[Tuple[int, StaleCause]]:
-        if cause.children is None:
-            yield (level, cause)
-        else:
-            for child in cause.children:
-                yield from self._gather_leaves(child, level=level + 1)
+        candidates = self._get_stale_causes(key=key)
```

Comment:
an alternative path that i couldn't quite reason through would be to prevent the cycles from being created in this `StaleCause` "tree"